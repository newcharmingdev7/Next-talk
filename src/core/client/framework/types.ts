// TODO: (@cvle) Extract useful common types into its own package.
export { Omit, Overwrite, PropTypesOf } from "coral-ui/types";
export { DeepPartial } from "coral-common/types";

export type RelayEnumLiteral<T> = keyof T | "%future added value";
