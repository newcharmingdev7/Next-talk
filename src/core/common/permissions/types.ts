export type UserRole =
  | "%future added value"
  | "COMMENTER"
  | "MEMBER"
  | "STAFF"
  | "MODERATOR"
  | "ADMIN";

export interface User {
  id: string;
  role: UserRole;
  moderationScopes?: null | {
    scoped?: boolean;
    siteIDs?: string[];
  };
}

export type Moderator = User & {
  role: "MODERATOR";
};

export type LTEModerator = User & {
  role: "MODERATOR" | "STAFF" | "MEMBER" | "COMMENTER";
};

export const isModerator = (user: User): user is Moderator => {
  return user.role === "MODERATOR";
};

export const isLTEModerator = (user: User): user is LTEModerator => {
  return user.role !== "ADMIN";
};
