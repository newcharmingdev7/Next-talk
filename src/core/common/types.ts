export type Diff<T extends keyof any, U extends keyof any> = ({ [P in T]: P } &
  { [P in U]: never } & { [x: string]: never })[T];

export type Omit<U, K extends keyof U> = Pick<U, Diff<keyof U, K>>;

export type Overwrite<T, U> = Pick<T, Diff<keyof T, keyof U>> & U;

export type Sub<T, U> = Pick<T, Diff<keyof T, keyof U>>;

/**
 * Make all properties in T writeable
 */
export type Writeable<T> = { -readonly [P in keyof T]: T[P] };

export type Promiseable<T> = Promise<T> | T;
