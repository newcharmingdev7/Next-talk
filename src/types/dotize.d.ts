declare module "dotize" {
  export function convert(obj: Object): Record<string, any>
}
