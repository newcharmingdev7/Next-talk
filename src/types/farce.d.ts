declare module "farce" {
  export class BrowserProtocol {}
  export class queryMiddleware {}
}
