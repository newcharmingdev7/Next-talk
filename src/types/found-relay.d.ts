declare module "found-relay" {
  import { Environment } from "relay-runtime";
  export class Resolver {
    constructor(relayEnvironment: Environment);
  }
}
