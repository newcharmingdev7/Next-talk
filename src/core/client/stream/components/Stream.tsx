import * as React from "react";
import { StatelessComponent } from "react";

import Logo from "talk-stream/components/Logo";
import CommentContainer from "talk-stream/containers/CommentContainer";
import PostCommentFormContainer from "talk-stream/containers/PostCommentFormContainer";
import { Button } from "talk-ui/components";

export interface StreamProps {
  id: string;
  isClosed: boolean;
  comments: ReadonlyArray<{ id: string }> | null;
  onLoadMore: () => void;
  hasMore: boolean;
}

const Stream: StatelessComponent<StreamProps> = props => {
  if (props.comments === null) {
    // TODO: (@cvle) What's the reason for this being null?
    return <div>Comments unavailable</div>;
  }
  return (
    <div>
      <Logo gutterBottom />
      <PostCommentFormContainer assetID={props.id} />
      {props.comments.map(comment => (
        <CommentContainer key={comment.id} data={comment} gutterBottom />
      ))}
      {props.hasMore && (
        <Button
          className={"talk-stream--loadmore"}
          onClick={props.onLoadMore}
          secondary
          invert
          fullWidth
        >
          Load More
        </Button>
      )}
    </div>
  );
};

export default Stream;
