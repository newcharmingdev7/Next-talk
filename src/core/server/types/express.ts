import { Request } from "express";

import { Tenant } from "talk-server/models/tenant";
import { User } from "talk-server/models/user";
import TenantCache from "talk-server/services/tenant/cache";

export interface Request extends Request {
  user?: User;
  tenant?: Tenant;
  tenantCache: TenantCache;
}
