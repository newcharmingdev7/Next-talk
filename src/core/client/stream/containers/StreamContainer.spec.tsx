import { shallow, ShallowWrapper } from "enzyme";
import { noop } from "lodash";
import React from "react";

import { PropTypesOf } from "talk-framework/types";

import Stream from "../components/Stream";
import { StreamContainer } from "./StreamContainer";

it("renders correctly", () => {
  const props: PropTypesOf<StreamContainer> = {
    asset: {
      id: "asset-id",
      isClosed: false,
      comments: {
        edges: [{ node: { id: "comment-1" } }, { node: { id: "comment-2" } }],
      },
    },
    relay: {
      hasMore: noop,
      isLoading: noop,
    } as any,
  };
  const wrapper = shallow(<StreamContainer {...props} />);
  expect(wrapper).toMatchSnapshot();
});

describe("when has more comments", () => {
  let finishLoading: ((error?: Error) => void) | null = null;
  const props: PropTypesOf<StreamContainer> = {
    asset: {
      id: "asset-id",
      isClosed: false,
      comments: {
        edges: [{ node: { id: "comment-1" } }, { node: { id: "comment-2" } }],
      },
    },
    relay: {
      hasMore: () => true,
      isLoading: () => false,
      loadMore: (_: any, callback: () => void) => (finishLoading = callback),
    } as any,
  };

  let wrapper: ShallowWrapper;

  beforeAll(() => (wrapper = shallow(<StreamContainer {...props} />)));

  it("renders hasMore", () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe("when loading more", () => {
    beforeAll(() => {
      wrapper
        .find(Stream)
        .props()
        .onLoadMore();
    });
    it("calls relay loadMore", () => {
      expect(finishLoading).not.toBeNull();
    });
    it("disables load more button", () => {
      wrapper.update();
      expect(wrapper).toMatchSnapshot();
    });
    it("enable load more button after loading is done", () => {
      finishLoading!();
      wrapper.update();
      expect(wrapper).toMatchSnapshot();
    });
  });
});
