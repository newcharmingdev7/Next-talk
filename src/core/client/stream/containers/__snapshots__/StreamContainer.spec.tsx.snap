// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Stream
  asset={
    Object {
      "comments": Object {
        "edges": Array [
          Object {
            "node": Object {
              "id": "comment-1",
            },
          },
          Object {
            "node": Object {
              "id": "comment-2",
            },
          },
        ],
      },
      "id": "asset-id",
      "isClosed": false,
    }
  }
  comments={
    Array [
      Object {
        "id": "comment-1",
      },
      Object {
        "id": "comment-2",
      },
    ]
  }
  disableLoadMore={false}
  onLoadMore={[Function]}
  user={null}
/>
`;

exports[`when has more comments renders hasMore 1`] = `
<Stream
  asset={
    Object {
      "comments": Object {
        "edges": Array [
          Object {
            "node": Object {
              "id": "comment-1",
            },
          },
          Object {
            "node": Object {
              "id": "comment-2",
            },
          },
        ],
      },
      "id": "asset-id",
      "isClosed": false,
    }
  }
  comments={
    Array [
      Object {
        "id": "comment-1",
      },
      Object {
        "id": "comment-2",
      },
    ]
  }
  disableLoadMore={false}
  hasMore={true}
  onLoadMore={[Function]}
  user={null}
/>
`;

exports[`when has more comments when loading more disables load more button 1`] = `
<Stream
  asset={
    Object {
      "comments": Object {
        "edges": Array [
          Object {
            "node": Object {
              "id": "comment-1",
            },
          },
          Object {
            "node": Object {
              "id": "comment-2",
            },
          },
        ],
      },
      "id": "asset-id",
      "isClosed": false,
    }
  }
  comments={
    Array [
      Object {
        "id": "comment-1",
      },
      Object {
        "id": "comment-2",
      },
    ]
  }
  disableLoadMore={true}
  hasMore={true}
  onLoadMore={[Function]}
  user={null}
/>
`;

exports[`when has more comments when loading more enable load more button after loading is done 1`] = `
<Stream
  asset={
    Object {
      "comments": Object {
        "edges": Array [
          Object {
            "node": Object {
              "id": "comment-1",
            },
          },
          Object {
            "node": Object {
              "id": "comment-2",
            },
          },
        ],
      },
      "id": "asset-id",
      "isClosed": false,
    }
  }
  comments={
    Array [
      Object {
        "id": "comment-1",
      },
      Object {
        "id": "comment-2",
      },
    ]
  }
  disableLoadMore={false}
  hasMore={true}
  onLoadMore={[Function]}
  user={null}
/>
`;
